### PR TITLE
perf(ivy): cache multiple reads to an element's stylingContext

### PR DESCRIPTION
**Pull request by matsko**
Because the styling context may be stored in a different location
and be apart of a sub array, reading the styling context each time
a host binding is evaluated is costly. This patch ensures that the
active styling context is cached so that multiple interactions with
styling bindings can easily retrieve the styling context efficiently.

### CODE REVIEW

File: packages/core/src/render3/instructions/styling.ts [hunk 1]
```diff
@@ -450,3 +443,12 @@ export function getActiveDirectiveStylingIndex() {
   // sub-classed directive the inheritance depth is taken into account as well.
   return getActiveDirectiveId() + getActiveDirectiveSuperClassDepth();
 }
+
+function getCachedStylingContext(index: number, lView?: LView) {
+  let context = getLastStoredStylingContext(index);
+  if (!context) {
+    context = getStylingContext(index + HEADER_OFFSET, lView || getLView());
```
Comment (kara) [inline packages/core/src/render3/instructions/styling.ts]:
```suggestion
    context = getStylingContextFromLView(index + HEADER_OFFSET, lView || getLView());
```

**Review by kara — CHANGES_REQUESTED** [review]
Like how small this PR is :-). Approach generally seems fine, but someone with more context on styling algorithm (@mhevery?) should probably vet it.

**Comment by matsko** [comment]
http://test/OCL:243321416:BASE:243344181:1555108125736:c5a06372

File: packages/core/src/render3/instructions/element.ts [hunk 2]
```diff
@@ -157,12 +157,12 @@ export function ɵɵelementEnd(): void {
...
   // context can be instantiated properly.
   if (hasClassInput(previousOrParentTNode)) {
-    const stylingContext = getStylingContext(previousOrParentTNode.index, lView);
+    const stylingContext = getStylingContextFromLView(previousOrParentTNode.index, lView);
     setInputsForProperty(
         lView, previousOrParentTNode.inputs !['class'] !, getInitialClassNameValue(stylingContext));
   }
   if (hasStyleInput(previousOrParentTNode)) {
-    const stylingContext = getStylingContext(previousOrParentTNode.index, lView);
+    const stylingContext = getStylingContextFromLView(previousOrParentTNode.index, lView);
```
Comment (kara) [inline packages/core/src/render3/instructions/element.ts]: Seems like we should cache access here too if it's expensive, aka 

```ts
let stylingContext: StylingContext;
if (hasClass...) {
  stylingContext = getStylingContextFrom LView(...);
}
if (hasStyle...) {
  stylingContext = stylingContext || getStylingContextFromLView(...)
}
```

**Review by kara — APPROVED** [review]
LGTM

File: packages/core/src/render3/instructions/styling.ts [hunk 3]
```diff
@@ -432,13 +425,22 @@ export function elementStylingApplyInternal(directiveStylingIndex: number, index
...
-  const stylingContext = getStylingContext(index + HEADER_OFFSET, lView);
+  const stylingContext = getStylingContext(index, lView);
   const totalPlayersQueued = renderStyling(
       stylingContext, renderer, lView, isFirstRender, null, null, directiveStylingIndex);
   if (totalPlayersQueued > 0) {
     const rootContext = getRootContext(lView);
     scheduleTick(rootContext, RootContextFlags.FlushPlayers);
   }
+
+  // because select(n) may not run between every instruction, the cached styling
```
Comment (mhevery) [inline packages/core/src/render3/instructions/styling.ts]: Select will get run between every instruction. Will that change the algorithm? Can you at  least change the comment.

File: packages/core/src/render3/instructions/styling.ts [hunk 4]
```diff
@@ -450,3 +452,12 @@ export function getActiveDirectiveStylingIndex() {
   // sub-classed directive the inheritance depth is taken into account as well.
   return getActiveDirectiveId() + getActiveDirectiveSuperClassDepth();
 }
+
+function getStylingContext(index: number, lView?: LView) {
```
Comment (mhevery) [inline packages/core/src/render3/instructions/styling.ts]:
```suggestion
function getStylingContext(index: number, lView: LView) {
```

File: packages/core/src/render3/instructions/styling.ts [hunk 5]
```diff
@@ -450,3 +452,12 @@ export function getActiveDirectiveStylingIndex() {
...
+  let context = getCachedStylingContext();
+  if (!context) {
+    context = getStylingContextFromLView(index + HEADER_OFFSET, lView || getLView());
```
Comment (mhevery) [inline packages/core/src/render3/instructions/styling.ts]: ### 
```suggestion
    context = getStylingContextFromLView(index + HEADER_OFFSET, lView);
```

File: packages/core/src/render3/instructions/styling.ts (hunk 4 above)
Comment (mhevery) [inline packages/core/src/render3/instructions/styling.ts]: Why not move this function into `state.ts`?

Comment (mhevery) [reply packages/core/src/render3/instructions/styling.ts]: also can you adde assertion which would run when `ngDevMode` is true which would assert that cached value is same as the looked up value.

**Review by mhevery — CHANGES_REQUESTED** [review]

**Review by mhevery — APPROVED** [review]

**Comment by matsko** [comment]
https://test.corp.google.com/ui#id=OCL:243321416:BASE:243670513:1555357865273:96a30cee

**Pull request closed without merging**